gli/core/flip.inl

```diff
@@ -45,27 +46,214 @@ namespace detail
 		}
 	}
 
-}//namespace detail
+	struct dxt1_block
+	{
+		uint16_t Color0;
+		uint16_t Color1;
+		uint8_t Row0;
+		uint8_t Row1;
+		uint8_t Row2;
+		uint8_t Row3;
+	};
 
-/*
-template <>
-inline image flip(image const & Image)
-{
+	struct dxt3_block
+	{
+		uint16_t AlphaRow0;
+		uint16_t AlphaRow1;
+		uint16_t AlphaRow2;
+		uint16_t AlphaRow3;
+		uint16_t Color0;
+		uint16_t Color1;
+		uint8_t Row0;
+		uint8_t Row1;
+		uint8_t Row2;
+		uint8_t Row3;
+	};
 
-}
-*/
+	struct dxt5_block
+	{
+		uint8_t Alpha0;
+		uint8_t Alpha1;
+		uint8_t AlphaR0;
+		uint8_t AlphaR1;
+		uint8_t AlphaR2;
+		uint8_t AlphaR3;
+		uint8_t AlphaR4;
+		uint8_t AlphaR5;
+		uint16_t Color0;
+		uint16_t Color1;
+		uint8_t Row0;
+		uint8_t Row1;
+		uint8_t Row2;
+		uint8_t Row3;
+	};
+	
+	inline void flip_block_s3tc(uint8_t* BlockDst, uint8_t* BlockSrc, format Format, bool HeightTwo)
+	{
+		// DXT1
+		if(Format == FORMAT_RGB_DXT1_UNORM || Format == FORMAT_RGB_DXT1_SRGB)
+		{
+			dxt1_block* Src = reinterpret_cast<dxt1_block*>(BlockSrc);
+			dxt1_block* Dst = reinterpret_cast<dxt1_block*>(BlockDst);
+	
+			if(HeightTwo)
+			{
+				Dst->Color0 = Src->Color0;
+				Dst->Color1 = Src->Color1;
+				Dst->Row0 = Src->Row1;
+				Dst->Row1 = Src->Row0;
+				Dst->Row2 = Src->Row2;
+				Dst->Row3 = Src->Row3;
+				
+				return;
+			}
+
+			Dst->Color0 = Src->Color0;
+			Dst->Color1 = Src->Color1;
+			Dst->Row0 = Src->Row3;
+			Dst->Row1 = Src->Row2;
+			Dst->Row2 = Src->Row1;
+			Dst->Row3 = Src->Row0;
+				
+			return;
+		}
+
+		// DXT1 w/ alpha
+		if(Format == FORMAT_RGBA_DXT1_UNORM || Format == FORMAT_RGBA_DXT1_SRGB)
+		{
+			// ?? can't find any spec for how a DXT1 texture with alpha is laid out
+			assert(false);
```

Comment:
There is no difference in the layout between `FORMAT_RGB_DXT1` and `FORMAT_RGBA_DXT1`. The only difference is the interpretation of the data.

When rendering in the RGB format the alpha bit results in a black RGB color with an alpha value of 1. With RGBA it results in a black RGB color with an alpha value of 0. Both formats counter-intuitively support alpha bits, they're only sampled differently.